// @bilims/mcp-sqlserver/dist/tools/list-views.js
import { BaseTool } from './base.js';
export class ListViewsTool extends BaseTool {
    getName() {
        return 'list_views';
    }
    getDescription() {
        return 'List all views in the current database or specified schema';
    }
    getInputSchema() {
        return {
            type: 'object',
            properties: {
                schema: {
                    type: 'string',
                    description: 'Schema name to filter views (optional)',
                },
            },
            required: [],
        };
    }
    async execute(params) {
        const { schema } = params;
        let query = `
      SELECT 
        TABLE_CATALOG as table_catalog,
        TABLE_SCHEMA as table_schema,
        TABLE_NAME as table_name,
        VIEW_DEFINITION as view_definition,
        CHECK_OPTION as check_option,
        IS_UPDATABLE as is_updatable
      FROM INFORMATION_SCHEMA.VIEWS
    `;
        if (schema) {
            query += ` WHERE TABLE_SCHEMA = '${schema.replace(/'/g, "''")}'`;
        }
        query += ' ORDER BY TABLE_SCHEMA, TABLE_NAME';
        return await this.executeSafeQuery(query);
    }
}
