// @bilims/mcp-sqlserver/dist/connection.js
import sql from 'mssql';
export class SqlServerConnection {
    pool = null;
    config;
    constructor(config) {
        this.config = config;
    }
    async connect() {
        if (this.pool) {
            return;
        }
        const sqlConfig = {
            server: this.config.server,
            database: this.config.database,
            user: this.config.user,
            password: this.config.password,
            port: this.config.port,
            options: {
                encrypt: this.config.encrypt,
                trustServerCertificate: this.config.trustServerCertificate,
                connectTimeout: this.config.connectionTimeout,
                requestTimeout: this.config.requestTimeout,
            },
            pool: {
                max: 10,
                min: 0,
                idleTimeoutMillis: 30000,
            },
        };
        this.pool = new sql.ConnectionPool(sqlConfig);
        await this.pool.connect();
    }
    async disconnect() {
        if (this.pool) {
            await this.pool.close();
            this.pool = null;
        }
    }
    async query(queryText) {
        if (!this.pool) {
            throw new Error('Database connection not established');
        }
        const request = this.pool.request();
        return await request.query(queryText);
    }
    async testConnection() {
        try {
            await this.connect();
            const result = await this.query('SELECT 1 as test');
            return result.recordset.length > 0;
        }
        catch (error) {
            return false;
        }
    }
    isConnected() {
        return this.pool !== null && this.pool.connected;
    }
    getConfig() {
        return { ...this.config };
    }
}
